File: app/src/main/java/com/trapero/cchoice/viewmodels/RegisterViewModel.java
package com.trapero.cchoice.viewmodels;

import android.util.Patterns;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.google.gson.JsonObject;
import com.trapero.cchoice.api.ApiService;
import com.trapero.cchoice.api.RetrofitClient;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RegisterViewModel extends ViewModel {

    private final MutableLiveData<String> fullName = new MutableLiveData<>();
    private final MutableLiveData<String> email = new MutableLiveData<>();
    private final MutableLiveData<String> password = new MutableLiveData<>();
    private final MutableLiveData<String> confirmPassword = new MutableLiveData<>();
    private final MutableLiveData<String> address = new MutableLiveData<>();
    private final MutableLiveData<String> contactNumber = new MutableLiveData<>();

    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private final MutableLiveData<Boolean> navigateToLogin = new MutableLiveData<>(false);
    private final MutableLiveData<Boolean> registrationSuccess = new MutableLiveData<>(false);
    private final MutableLiveData<Boolean> isRegistering = new MutableLiveData<>(false); // To show loading state

    public LiveData<String> getFullName() {
        return fullName;
    }

    public LiveData<String> getEmail() {
        return email;
    }

    public LiveData<String> getPassword() {
        return password;
    }

    public LiveData<String> getConfirmPassword() {
        return confirmPassword;
    }

    public LiveData<String> getAddress() {
        return address;
    }

    public LiveData<String> getContactNumber() {
        return contactNumber;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<Boolean> getNavigateToLogin() {
        return navigateToLogin;
    }

    public LiveData<Boolean> getRegistrationSuccess() {
        return registrationSuccess;
    }

    public LiveData<Boolean> getIsRegistering() {
        return isRegistering;
    }

    public void onSignUpClicked() {
        String fullNameValue = fullName.getValue();
        String emailValue = email.getValue();
        String passwordValue = password.getValue();
        String confirmPasswordValue = confirmPassword.getValue();
        String addressValue = address.getValue();
        String contactNumberValue = contactNumber.getValue();

        if (fullNameValue == null || fullNameValue.trim().isEmpty()) {
            errorMessage.setValue("Please enter your full name.");
            return;
        }
        if (emailValue == null || !Patterns.EMAIL_ADDRESS.matcher(emailValue).matches()) {
            errorMessage.setValue("Please enter a valid email address.");
            return;
        }
        if (passwordValue == null || passwordValue.length() < 6) {
            errorMessage.setValue("Password must be at least 6 characters long.");
            return;
        }
        if (confirmPasswordValue == null || !confirmPasswordValue.equals(passwordValue)) {
            errorMessage.setValue("Passwords do not match.");
            return;
        }

        isRegistering.setValue(true);
        JsonObject registrationData = new JsonObject();
        registrationData.addProperty("fullName", fullNameValue);
        registrationData.addProperty("email", emailValue);
        registrationData.addProperty("password", passwordValue);
        registrationData.addProperty("password_confirmation", confirmPasswordValue);
        registrationData.addProperty("address", addressValue);
        registrationData.addProperty("contactNumber", contactNumberValue);

        ApiService apiService = RetrofitClient.getApiService();
        Call<JsonObject> call = apiService.registerUser(registrationData);
        call.enqueue(new Callback<JsonObject>() {
            @Override
            public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                isRegistering.setValue(false);
                if (response.isSuccessful() && response.body() != null) {
                    // Handle successful registration response
                    String message = response.body().get("message").getAsString();
                    registrationSuccess.setValue(true);
                } else {
                    // Handle unsuccessful registration response (e.g., validation errors)
                    if (response.errorBody() != null) {
                        try {
                            String errorString = response.errorBody().string();
                            // You might want to parse this JSON error to display specific messages
                            errorMessage.setValue("Registration failed: " + errorString);
                        } catch (Exception e) {
                            errorMessage.setValue("Registration failed: An unknown error occurred.");
                        }
                    } else {
                        errorMessage.setValue("Registration failed: No error body.");
                    }
                }
            }

            @Override
            public void onFailure(Call<JsonObject> call, Throwable t) {
                isRegistering.setValue(false);
                errorMessage.setValue("Network error: " + t.getMessage());
            }
        });
    }

    public void onLoginClicked() {
        navigateToLogin.setValue(true);
    }

    public void doneNavigatingToLogin() {
        navigateToLogin.setValue(false);
    }

    public void doneRegistering() {
        registrationSuccess.setValue(false);
    }

    public void doneShowingError() {
        errorMessage.setValue(null);
    }
}

File: app/src/main/java/com/trapero/cchoice/viewmodels/RegisterViewModel.kt
package com.trapero.cchoice.viewmodels

import android.util.Log
import android.util.Patterns
import android.view.View
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.reflect.TypeToken
import com.trapero.cchoice.api.ApiHelper
import com.trapero.cchoice.api.RetrofitClient
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

class RegisterViewModel : ViewModel() {

    private val _firstName = MutableLiveData<String>()
    var firstName: MutableLiveData<String>
        get() = _firstName
        set(value) { _firstName.value = value.value }

    private val _lastName = MutableLiveData<String>()
    var lastName: MutableLiveData<String>
        get() = _lastName
        set(value) { _lastName.value = value.value }

    private val _email = MutableLiveData<String>()
    var email: MutableLiveData<String>
        get() = _email
        set(value) { _email.value = value.value }

    private val _password = MutableLiveData<String>()
    var password: MutableLiveData<String>
        get() = _password
        set(value) { _password.value = value.value }

    private val _confirmPassword = MutableLiveData<String>()
    var confirmPassword: MutableLiveData<String>
        get() = _confirmPassword
        set(value) { _confirmPassword.value = value.value }

    private val _address = MutableLiveData<String>()
    var address: MutableLiveData<String>
        get() = _address
        set(value) { _address.value = value.value }

    private val _contactNumber = MutableLiveData<String>()
    var contactNumber: MutableLiveData<String>
        get() = _contactNumber
        set(value) { _contactNumber.value = value.value }

    private val _errorMessage = MutableLiveData<String?>()
    var errorMessage: MutableLiveData<String?>
        get() = _errorMessage
        set(value) { _errorMessage.value = value.value }

    private val _navigateToLogin = MutableLiveData(false)
    var navigateToLogin: MutableLiveData<Boolean>
        get() = _navigateToLogin
        set(value) { _navigateToLogin.value = value.value }

    private val _registrationSuccess = MutableLiveData(false)
    var registrationSuccess: MutableLiveData<Boolean>
        get() = _registrationSuccess
        set(value) { _registrationSuccess.value = value.value }

    private val _isRegistering = MutableLiveData(false)
    var isRegistering: MutableLiveData<Boolean>
        get() = _isRegistering
        set(value) { _isRegistering.value = value.value }


    val signUpClickedListener = View.OnClickListener {
        onSignUpClicked()
    }

    fun onSignUpClicked() {

        isRegistering.value = true


        val apiHelper = ApiHelper()

        apiHelper.setBody<JsonObject>(mapOf(
            "user_first_name" to firstName.value.toString(),
            "user_last_name" to lastName.value.toString(),
            "user_email" to email.value.toString(),
            "user_password" to password.value.toString(),
            "user_phone" to contactNumber.value.toString(),
            "user_address" to address.value.toString(),
            "user_type" to 0
        ))
        .post("/register", object : okhttp3.Callback {
            override fun onFailure(call: okhttp3.Call, e: IOException) {
                isRegistering.postValue(false)
                errorMessage.postValue(e.message)
            }

            override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
                isRegistering.postValue(false)
                if (response.isSuccessful) {
                    registrationSuccess.postValue(true)
                } else {
                    val mapType = object : TypeToken<Map<String, Any>>() {}.type
                    val responseMap: Map<String, Any> = Gson().fromJson(response.body?.string(), mapType)
                    val errors = responseMap["errors"] as? Map<String, ArrayList<String>>
                    val firstErrorMessage = errors?.values?.firstOrNull()?.firstOrNull()

                    firstErrorMessage?.let {
                        errorMessage.postValue(it)  // Post the first error message
                    } ?: errorMessage.postValue("No error found")

                }
            }
        })
    }

    fun onLoginClicked() {
        navigateToLogin.value = true
    }

    fun doneNavigatingToLogin() {
        navigateToLogin.value = false
    }

    fun doneRegistering() {
        registrationSuccess.value = false
    }

    fun doneShowingError() {
        errorMessage.value = null
    }
}
